fluent/syntax/ast.py

```diff
@@ -103,18 +103,12 @@ def equals(self, other, ignored_fields=['span']):
                 if len(field1) != len(field2):
                     return False
 
-                # These functions are used to sort lists of items for when
-                # order doesn't matter.  Annotations are also lists but they
-                # can't be keyed on any of their fields reliably.
-                field_sorting = {
-                    'attributes': lambda elem: elem.id.name,
-                    'variants': lambda elem: elem.key.name,
-                }
-
-                if key in field_sorting:
-                    sorting = field_sorting[key]
-                    field1 = sorted(field1, key=sorting)
-                    field2 = sorted(field2, key=sorting)
+                # Sort elements of order-agnostic fields to ensure the
+                # comparison is order-agnostic as well. Annotations should be
+                # here too but they don't have sorting keys.
+                if key in ('attributes', 'variants'):
+                    field1 = sorted(field1, key=lambda elem: elem.sorting_key)
+                    field2 = sorted(field2, key=lambda elem: elem.sorting_key)
```

Comment:
This isn't very DRY, but at least it's readable. I considered using (`attrgetter`](https://docs.python.org/3.6/library/operator.html#operator.attrgetter) but in the end I thought it would have hurt readability too.